### PR TITLE
Add Metadata And Ping RPC methods

### PR DESCRIPTION
**Pull request by nisdas**
Part of #5227 and builds on https://github.com/ethereum/eth2.0-specs/pull/1684 from the eth2 spec.

- [x] Adds Ping RPC Method
- [x] Adds Metatdata RPC method.
- [x] Add flag and ability to persist node metadata to disk.
- [x] Refactor rpc topic mappings to simply refer to topic constants instead.
- [x] Add tests for new RPC methods.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Any reason this can't go into master?

**Comment by nisdas** [comment]
@prestonvanloon it can't, it is v0.11 specific. The previous versions of the spec have no use for these rpc methods so it wouldn't make sense to put this in master

File: beacon-chain/sync/rpc_metadata_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+
+	"github.com/prysmaticlabs/go-ssz"
+
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_metadata_test.go:14]:
```suggestion
```

File: beacon-chain/sync/rpc_metadata_test.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+
+	"github.com/prysmaticlabs/go-ssz"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_metadata_test.go:12]:
```suggestion
```

File: beacon-chain/sync/rpc_metadata_test.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
+package sync
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_metadata_test.go:10]:
```suggestion
```

File: beacon-chain/sync/rpc_ping.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
+package sync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_ping.go:10]:
```suggestion
```

File: beacon-chain/sync/rpc_ping.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/libp2p/go-libp2p-core/peer"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_ping.go:12]:
```suggestion
```

File: beacon-chain/sync/rpc_ping_test.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
+package sync
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_ping_test.go:10]:
```suggestion
```

File: beacon-chain/sync/rpc_status.go [hunk 7]
```diff
@@ -5,6 +5,8 @@ import (
 	"context"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+
```
Comment (nisdas) [inline beacon-chain/sync/rpc_status.go:9]:
```suggestion
```

File: beacon-chain/p2p/interfaces.go [hunk 8]
```diff
@@ -3,6 +3,8 @@ package p2p
 import (
 	"context"
 
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+
```
Comment (nisdas) [inline beacon-chain/p2p/interfaces.go:7]:
```suggestion
```

**Comment by terencechain** [comment]
@nisdas linter failed

**Comment by terencechain** [comment]
can you also run goimports please?

**Comment by nisdas** [comment]
@terencechain ptal, made the changes

File: beacon-chain/p2p/service.go [hunk 9]
```diff
@@ -327,6 +335,16 @@ func (s *Service) Peers() *peers.Status {
 	return s.peers
 }
 
+// Metadata returns a copy of the peer's metadata.
+func (s *Service) Metadata() *pb.MetaData {
+	return proto.Clone(s.metaData).(*pb.MetaData)
+}
+
+// MetadataSeq returns the metadata sequence number
```
Comment (terencechain) [inline beacon-chain/p2p/service.go:343]:
```suggestion
// MetadataSeq returns the metadata sequence number.
```

File: shared/cmd/flags.go [hunk 10]
```diff
@@ -112,6 +112,12 @@ var (
 		Usage: "The file containing the private key to use in communications with other peers.",
 		Value: "",
 	}
+	// P2PMetadata defines a flag to specify the location of the peer metadata file.
+	P2PMetadata = &cli.StringFlag{
+		Name:  "p2p-metaData",
```
Comment (terencechain) [inline shared/cmd/flags.go:117]:
```suggestion
		Name:  "p2p-metadata",
```

**Pull request merged**